gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/iap/IAPAuthenticationFilter.groovy

```diff
@@ -0,0 +1,204 @@
+
+  PermissionService permissionService
+
+  Front50Service front50Service
+
+  private static final String signatureAttribute = "JWTSignature"
+
+  private final Map<String, JWK> keyCache = new HashMap<>()
+  private final Logger logger = LoggerFactory.getLogger(IAPAuthenticationFilter.class)
+
+  IAPAuthenticationFilter(
+    IAPConfig.IAPSecurityConfigProperties configProperties,
+    PermissionService permissionService,
+    Front50Service front50Service) {
+    this.configProperties = configProperties
+    this.permissionService = permissionService
+    this.front50Service = front50Service
+  }
+
+  @Override
+  void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+    HttpServletRequest req = (HttpServletRequest) request
+    HttpServletResponse res = (HttpServletResponse) response
+
+
+    try {
+      def token = req.getHeader(configProperties.jwtHeader)
+      Preconditions.checkNotNull(token)
+
+      SignedJWT jwt = SignedJWT.parse(token)
+
+
+      def signatureInSession = req.getSession(false)?.getAttribute(signatureAttribute)
+
+
+      if (signatureInSession && signatureInSession == jwt.signature) {
+        // Signature matches in previous request signatures in current session, skip validation.
+        chain.doFilter(req, res)
+        return
+      }
+
+      def verifiedUser = verifyJWTAndGetUser(jwt)
+
+      PreAuthenticatedAuthenticationToken authentication = new PreAuthenticatedAuthenticationToken(
+        verifiedUser,
+        null /* credentials */,
+        verifiedUser.authorities
+      )
+
+      // Service accounts are already logged in.
+      if (!isServiceAccount(verifiedUser.email)) {
+        permissionService.login(verifiedUser.email)
+      }
+
+      SecurityContextHolder.getContext().setAuthentication(authentication)
+
+      // Save the signature to skip validation for subsequent requests with same token.
+      req.getSession().setAttribute(signatureAttribute, jwt.signature)
+
+    } catch (Exception e) {
+      logger.error("Could not verify JWT Token", e)
+      res.sendError(HttpServletResponse.SC_FORBIDDEN, "Could not verify JWT Token")
+      req.getSession(false)?.invalidate()
+      return
+    }
+
+    chain.doFilter(req, res)
+  }
+
+  @Override
+  void init(FilterConfig filterConfig) throws ServletException {}
+
+  @Override
+  void destroy() {}
+
+  boolean isServiceAccount(String email) {
+    if (!email || !permissionService.isEnabled()) {
+      return false
+    }
+    try {
+      def serviceAccounts = front50Service.getServiceAccounts()
+      return serviceAccounts.find { email.equalsIgnoreCase(it.name) }
+    } catch (RetrofitError re) {
+      logger.warn("Could not get list of service accounts.", re)
+    }
+    return false
+  }
+
+  private User verifyJWTAndGetUser(SignedJWT jwt) throws Exception {
+    JWSHeader jwsHeader = jwt.getHeader()
+
+    // Header must have algorithm("alg") and "kid"
```

Comment:
I'm of the opinions that comments should describe _why_ code is needed or structured in a certain way, and not to restate _what_ it's doing (just in English). Applies to all comments below.

Reply:
Removed.